src/InboxMonitor.vala

```diff
@@ -0,0 +1,228 @@
+        }
+    }
+
+    private async void inbox_folder_synchronize_sync (E.Source source) {
+        if (!network_monitor.network_available) {
+            debug ("[%s] Network is not avaible. Skipping…", source.display_name);
+            return;
+        }
+
+        var inbox_folder = inbox_folders.get (source);
+        if (inbox_folder != null) {
+            debug ("[%s] Refreshing…", source.display_name);
+
+            try {
+                inbox_folder.refresh_info_sync (null);
+
+            } catch (Error e) {
+                warning ("[%s] Error refreshing: %s", source.display_name, e.message);
+            }
+        }
+    }
+
+    private void inbox_folder_changed (E.Source source, Camel.FolderChangeInfo changes) {
+        var inbox_folder = inbox_folders.get (source);
+        if (inbox_folder == null) {
+            return;
+        }
+
+        unowned var added_uids = changes.get_added_uids ();
+        if (added_uids != null) {
+            var sender_names = new GenericSet<string> (str_hash, str_equal);
+            var unseen_message_infos = new SList<Camel.MessageInfo> ();
+
+            added_uids.foreach ((added_uid) => {
+                var message_info = inbox_folder.get_message_info (added_uid);
+
+                if (!(Camel.MessageFlags.SEEN in message_info.flags)) {
+                    unowned string? sender_address;
+                    unowned string? sender_name;
+
+                    var camel_address = new Camel.InternetAddress ();
+                    camel_address.unformat (message_info.from);
+                    camel_address.get (0, out sender_name, out sender_address);
+
+                    if (sender_name == null) {
+                        sender_name = sender_address;
+                    }
+
+                    sender_names.add (sender_name);
+                    unseen_message_infos.append (message_info);
+                }
+            });
+
+            var unseen_message_infos_length = unseen_message_infos.length ();
+            if (unseen_message_infos_length == 1) {
+                var unseen_message_info = unseen_message_infos.nth_data (0);
+
+                var notification = new GLib.Notification (sender_names.iterator ().next_value ());
+                notification.set_body (unseen_message_info.subject);
+                GLib.Application.get_default ().send_notification (unseen_message_info.uid, notification);
+
+            } else if (unseen_message_infos_length > 1) {
+                GLib.Notification notification;
+
+                if (sender_names.length == 1) {
+                    notification = new GLib.Notification (_("%u new messages from %s").printf (unseen_message_infos_length, sender_names.iterator ().next_value ()));
+                } else {
+                    notification = new GLib.Notification (_("%u new messages from %u senders").printf (unseen_message_infos_length, sender_names.length));
+                }
```

Comment:
I guess we need to use ngettext for languages that have multiple plural forms, maybe something like this?

```vala
GLib.Notification notification;
string messages_count = ngettext ("%u new message", "%u new messages", unseen_message_infos_length).printf (unseen_message_infos_length);

if (sender_names.length == 1) {
    ///TRANSLATORS: The first %s represents the number of new messages translated in your language, e.g. "2 new messages"
    ///The next %s represents the sender name
    notification = new GLib.Notification (_("%s from %s").printf (messages_count, sender_names.iterator ().next_value ()));
} else {
    ///TRANSLATORS: The first %s represents the number of new messages translated in your language, e.g. "2 new messages"
    ///The next %s represents the number of senders
    notification = new GLib.Notification (ngettext ("%s from %u sender", "%s from %u senders", sender_names.length).printf (messages_count, sender_names.length));
}
```

Reply:
Thanks - that's really good to know! I struggled exactly with this - but then found we can turn the bug into a feature:

If all the messages are from the same sender, it would be nice to know from whom. So instead of simply displaying "3 messages from 1 sender" we are then able to display "3 messages from Ryo Nakano".

Which I think does provide more value because if I see something like "10 messages from 1 sender" I'm probably interested who this guy is anyway 😆 

What do you think, @danrabbit?

Reply:
... unless I'm missing something here? Because of the way the notifications are arranged, I thought we don't need the plural/singular distinction...?

Reply:
Yeah you still need to use ngettext for languages with multiple plural forms, but I agree that if there's only one sender it makes sense to call out that sender specifically rather than say `1`